### PR TITLE
QWERTY keyboard piano trigger support

### LINKED ISSUE #20: Add qwerty keyboard piano trigger support

### LINKED ISSUE #28: rewrite Piano UI
the way we're placing the piano keys is quite messed up and the current key bindings only apply to the first octave which is so low in frequency and doesn't sound good to the ears.

```svelte
<div
    on:mousedown={handleKeyDown}
    on:mouseenter={handleKeyEnter}
    on:focus
    style={applyTheme(vars) +
        `${black ? `left: ${keySpacing * 2.2 + 1 + 2 * 0.2}rem;` : ''}`}
    class="key"
    class:black
    class:pressed={$PianoStore.currNote === key}>
    <p>{key.note.name[0]}</p>
</div>
```

image

### PR DESCRIPTION
**Pull request by harveysanders**
Resolves #20 
Resolves #28 

Add keybindings for triggering piano

TODO:
- [ ] Add octave shift binding support
- [ ] Add settings for changing the piano key labels (show key binding, vs note name, vs MIDI number)

image

### CODE REVIEW

File: src/lib/components/jam/PianoKey.svelte [hunk 1]
```diff
@@ -1,103 +1,118 @@
...
+		}
 
-        $PianoStore.currNote = key;
-    }
+		$PianoStore.currNote = key;
+	}
 
-    let vars;
-    $: vars = $themeStore.vars;
+	function keyLabel(labelType: KeyLabel, key: PianoKeyNote): string {
```
Comment (harveysanders) [inline src/lib/components/jam/PianoKey.svelte:42]: The main code is here. The rest is from prettier

File: src/lib/components/jam/PianoKey.svelte [hunk 2]
```diff
@@ -1,103 +1,118 @@
...
+	on:mousedown={handlePress}
+	on:mouseenter={handleKeyEnter}
+	on:focus
+	style={applyTheme(vars) +
+		`${black ? `left: ${keySpacing * 2.2 + 1 + 2 * 0.2}rem;` : ""}`}
+	class="key"
+	class:black
+	class:pressed={$PianoStore.currNote === key}>
+	<!-- TODO: Switch label type in settings -->
+	<p>{keyLabel("binding", key)}</p>
```
Comment (harveysanders) [inline src/lib/components/jam/PianoKey.svelte:69]: `keyLabel` usage here

File: src/lib/components/global/TextInput.svelte [hunk 3]
```diff
@@ -15,6 +15,7 @@
     {placeholder}
     {disabled}
     on:input
+    on:keydown|stopPropagation
```
Comment (harveysanders) [inline src/lib/components/global/TextInput.svelte:18]: Do we want to do this for all `TextInput`s or should I add a flag prop to toggle `stopPropagation` on events on specific instances of `TextInput`?

Comment (pmoieni) [reply src/lib/components/global/TextInput.svelte:18]: If this resolves the piano playing while typing issue then yes do it for all components

Comment (harveysanders) [reply src/lib/components/global/TextInput.svelte:18]: Yeah the `stopPropagation` event method stops the event from bubbling past the call, so in this case this prevents `keydown` events from bubbling up to the window.  
https://developer.mozilla.org/en-US/docs/Web/API/Event/stopPropagation

**Comment by adoublef** [comment]
regarding the UI, I think the idea to only display keys they user will have access to with bindings is smart

- black keys `wetyuop` 
-  white keys `asdfghjkl;'`
- down/up octave `-+` is a bit more intuitive and easier to press than the suggested `zx` key pairing

**Comment by harveysanders** [comment]
Switched to new branch to resolve merge conflicts more cleanly:
https://github.com/Rapidmidiex/rmx-web-client/pull/34

**Pull request closed without merging**